File: 0003-Longest-Substring-Without-Repeating-Characters/lc0003.py
# Python solution.

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        poslog = {}
        start, ans = (0, 0)
        for cursor, val in enumerate(s):
            query_pos = poslog.get(val)
            start = max(start, query_pos if query_pos else 0)
            ans = max(ans, cursor-start+1)
            poslog[val] = cursor+1
        return ans

File: 0003-Longest-Substring-Without-Repeating-Characters/lc0003.js
// JavaScript solution.

/**
 * @param {string} s
 * @return {number}
 */

var lengthOfLongestSubstring = function(s) {
    let poslog = new Map();
    let start = 0, ans = 0;
    for (const [cursor, val] of [...s].entries()) {
        query_pos = poslog.get(val);
        if (query_pos !== undefined && query_pos > start) {
            start = query_pos;
        }
        ans = Math.max(ans, cursor-start+1);
        poslog.set(val, cursor+1);
    }
    return ans;
};
